### ISSUE TITLE
What could cause event propagation to fail?

### ISSUE THREAD

**Issue by kaansoral** [pixijs/pixi.js#2921]
Hi

I've noticed a propagation issue today, I've been debugging it for hours, it has consumed me, I've inspected the PIXI source, tested various things, but no luck

I've tested various PIXI versions, including the latest dev, they also respond the same. I've also tried to reproduce the issue, however I couldn't

I suspect pixi-display could be causing the issue, but it seems unlikely - however, since the interaction code is complicated (I couldn't grasp the bubbling logic in a short span of time), it might be so

**The issue:** There is a character, under the character, there is a monster, I right click the character, the event isn't stopped, yet the right click doesn't bubble down, .stopped is false, the event just doesn't bubble :(

I've tested a normal click too, no matter what I do, I can't make the event bubble once a callback is fired (so basically, normally, when the event.stopPropagation isn't called, the event should bubble down on it's own)

**Comment by ivanpopelyshev** [comment]
That's not how its bubbling. I thought its bubbling into the parent? You can't just ask character to ignore the event. If its interactive, it will receive all events, and stop his "brothers" from doing that.

**Comment by kaansoral** [comment]
Very basically, there are 2 objects on the map

When one object overlaps the other, I want the event to bubble down, it doesn't bubble down

They are on the same displayGroup, they are different/separate entities

I just managed to reproduce the issue by the way, simply, when you overlap 2 Sprite's that are on the same displayList, the event bubbles down to the map, yet the event doesn't trigger on both Sprite's, as it should trigger

**Comment by ivanpopelyshev** [comment]
Can you prepare an example that we can put to http://pixijs.github.io/examples/index.html?s=basics&f=container.js&title=Container&v=dev#/basics/basic.js ?

**Comment by kaansoral** [comment]
For example, on this example: http://pixijs.github.io/examples/index.html?s=basics&f=container.js&title=Container&v=dev#/display/zorder.js

There is no `event.stopPropagation` on `onDragStart` - yet the event stops, I think, when you drag(or specifically, click) something on an overlapping point, you should be dragging multiple entities at the same time

**Comment by ivanpopelyshev** [comment]
event bubbles to parents, not to brothers. Its not possible to drag everything :( @GoodBoyDigital any ideas on that?

**Comment by kaansoral** [comment]
Well that clears things up, thank you, as I suspected, it wasn't a pixi-display issue after all, just re-reproduced it without pixi-display, and as you said, there is no reproduction needed, since I wanted/expected sibling propagation :)

My specific use case is, on an mmorpg, when one player overlaps a monster, you should hit monster, when you click the overlapping player

I had a very simple and elegant logic that assumed propagation worked between siblings too, now, I will probably move to a much complicated system that adds/removes listeners depending on things

For example, in a PVP region, other players should have attack handlers, but on a regular region, players shouldn't have attack handlers, and monsters should absorb the click event - if sibling propagation worked, the code was super simple and elegant - but very luckily, at least there is a complex solution out there for me

Still, I think people might need/expect sibling propagation, so please @GoodBoyDigital :)

**Comment by ivanpopelyshev** [comment]
If you want to add sibling propagation, feel free to clone pixi-display and change https://github.com/pixijs/pixi-display/blob/master/src/InteractionManagerMixin.js that way queue is sorted array of all things under the mouse

**Comment by kaansoral** [comment]
Your proposed solution initially seemed complex to me, so I decided to try the conditional listener approach

However, I quickly realised that (not as quickly as I wanted, I realised after the solution was implemented)

**If a sprite is .interactive, it catches all events, not just the listened ones.** I hoped/assumed the logic would be selective

Checked `InteractionManagerMixin` after this experience, it took me quite a while to understand how it integrates

I'm looking for an easy way to refactor `.interactive` to `.interactive && this._events[evt]` in the logic, and add that as a mixin, does it make sense?

**Comment by ivanpopelyshev** [comment]
Yep, that makes sence :)

**Comment by kaansoral** [comment]
In addition, a `.continuePropagation` could be nice too

It seems I didn't know how exactly DOM propagation worked, as I inspected the DOM propagation, PIXI emulates DOM events almost 1:1 - non .interactive elements act like `pointer-events: none`

So what I'm proposing/needing is extending the DOM events, but they could be really useful and practical for games

1) `.interactive + handler check` - this can be enabled by an InteractionManager argument
2) `.continuePropagation` - this could set a variable at the event, so the "hit" state is reversed, and `.processInteractive` can continue to check for hits
At least I hope

None of these routines support these additions without an arguments refactoring, so, I will check now whether they can be hacked in by setting flags at objects, and can be mixed in the way you mixed in the pixi-display stuff

**Comment by ivanpopelyshev** [comment]
It's difficult to make. The thing is, processInteractive first sorts everything and then decides which element was hit. You have to implmenet my approach first, then this method will be available.

**Comment by kaansoral** [comment]
Do you think it's possible by just modifying `_finishInteractionProcess` of your Mixin?

When a flag is set, it could sort the queue[0] and re-call the `func`, until that specific flag isn't set again

**Comment by ivanpopelyshev** [comment]
yeah, just sort it by displayOrder. queue[1] has stuff that hit, while queue[0] are rejects.

Also look at that line: https://github.com/pixijs/pixi-display/blob/master/src/InteractionManagerMixin.js#L74 . It actually doesnt process stuff that is worse than current queue[1], you have to modify it too

**Comment by kaansoral** [comment]
Ah, thanks, in that case, since we are in the custom code zone now, probably a `.missable` property is a good idea too, to only partially lose performance by bypassing that line - if the best hit is missable, it should keep on collecting things

**Comment by ivanpopelyshev** [comment]
Yes, that's good idea.

**Comment by englercj** [comment]
To clarify for myself later: This request is to add a feature when an interaction event can go "through" an interactive object if that object decides not to handle the event.

**Comment by staff0rd** [comment]
I just ran into this and want to point out it's specific to pixi, not pixi-display.

In the following, you cannot click the blue box because it's inside a parent that is flagged interactive.  https://codepen.io/staff0rd/pen/pEVmEk?editors=0010

![image](https://cloud.githubusercontent.com/assets/6289998/19300330/74c52e92-90a5-11e6-8fa9-7388bb95f058.png)

**Comment by kaansoral** [comment]
Interestingly, while I didn't inspect your example in detail as I don't know the addChild(a,b) format, your example works as I wanted event propagation to work

If you add an alert to child and parent click events, first the child gets called, then the parent

**Comment by staff0rd** [comment]
Indeed you are right.  I thought this was reproducing an issue I had found in a project, however even including pixi-display in the pen still has the child being hit.  I'll have to look further into the project.

**Comment by kaansoral** [comment]
`event.stopPropagation();` seems to solve the problem/challenge in your codepen example

**Comment by billschaller** [comment]
Here's a good example of the sibling problem:

https://codepen.io/anon/pen/yaWVoR

When you click the area where the sibling is overlapping the child block, the click doesn't register through to the child block, even though there is no click event even registered on the sibling block. 

That does make sense in context with the DOM model, but in the case of game interactions, there are many scenarios where you would want overlapping interactives, especially, for instance, overlapping interactives hooking different events. A good example would be an RTS selection mechanism or a 2d graph brushing mechanism.

**Comment by staff0rd** [comment]
Thanks @billschaller, here's that pen pointing to latest; https://codepen.io/staff0rd/pen/WoYqam

So correct me if I'm wrong, but this issue is about allowing interaction events to fall through to siblings of a parent that have a lower z-order than the front one that originally picked it up?

Also related to the remaining item in #1901, although that one is about a non-interactive sibling blocking an interactive (behind) sibling which apparently didn't occur in v2.

**Comment by englercj** [comment]
Is this still an issue?

**Comment by LFP6** [comment]
Still interested in the ability to let events pass through to siblings, as far as I know there's no simple way to do this currently.

**Comment by LFP6** [comment]
Still interested in the ability to let events pass through to siblings, as far as I know there's no simple way to do this currently.

**Comment by themoonrat** [comment]
How about something like this. It's not fit for purpose, but as a proof of concept.....
https://www.pixiplayground.com/#/edit/~GWJRGg1Tnx~hEh6dnd01

On the down event of the front bunny, I make it not interactive, then restart the interaction processing on the parent, allowing the next thing behind it to grab an interaction.

Would be quite easy to turn this into a util function where required.
